File: 2019/java/Advent22.java
import java.math.*;
import java.util.*;
import java.util.stream.*;

public class Advent22 extends Advent {

  //int DECK_SIZE = 10;
  int DECK_SIZE = 10_007;

  List<Integer> cards;

  public Advent22() {
    super(22);
  }

  @Override
  protected void parseInput() {
    cards = IntStream.range(0,DECK_SIZE).boxed().collect(Collectors.toList());
  }

  @Override
  protected Object part1() {
    readInstructions();
    return cards.indexOf(2019);
  }

  long mod(long n) {
    return mod(n, DECK_SIZE);
  }

  long mod(long n, long k) {
    long result = n % k;
    return result < 0 ? result + k : result;
  }

  void readInstructions() {
    for (String line : input) {
      String[] ss = line.split(" ");
      if (ss[1].equals("into")) {
        dealIntoNewStack();
        continue;
      }
      if (ss[1].equals("with")) {
        dealWithIncrement(Integer.parseInt(ss[3]));
        continue;
      }
      if (ss[0].equals("cut")) {
        cut(Integer.parseInt(ss[1]));
        continue;
      }
    }
  }

  void dealIntoNewStack() {
    Collections.reverse(cards);
  }

  void dealWithIncrement(int n) {
    List<Integer> newCards = IntStream.range(0,DECK_SIZE).boxed().collect(Collectors.toList());
    for (int i = 0; i < DECK_SIZE; i++) {
      newCards.set((i*n)%DECK_SIZE, cards.get(i));
    }
    cards = newCards;
  }

  void cut(int n) {
    List<Integer> newCards = new ArrayList<>(DECK_SIZE);
    if (n < 0) {
      newCards.addAll(cards.subList(DECK_SIZE + n, DECK_SIZE));
      newCards.addAll(cards.subList(0, DECK_SIZE + n));
    } else {
      newCards.addAll(cards.subList(n, DECK_SIZE));
      newCards.addAll(cards.subList(0, n));
    }
    cards = newCards;
  }

  @Override
  protected Object part2() {
    BigInteger size = BigInteger.valueOf(119_315_717_514_047L);
    BigInteger shuffles = BigInteger.valueOf(101_741_582_076_661L);
    BigInteger skip = BigInteger.ONE;
    BigInteger start = BigInteger.ZERO;
    var reverse = new ArrayList<>(input);
    Collections.reverse(reverse);
    for (String line : reverse) {
      String[] ss = line.split(" ");
      if (ss[1].equals("into")) {
        skip = skip.negate().mod(size);
        start = start.negate().add(size).subtract(BigInteger.ONE).mod(size);
        continue;
      }
      if (ss[1].equals("with")) {
        BigInteger inc = new BigInteger(ss[3]);
        skip = skip.multiply(inc.modInverse(size)).mod(size);
        start = start.multiply(inc.modInverse(size)).mod(size);
        continue;
      }
      if (ss[0].equals("cut")) {
        BigInteger amt = new BigInteger(ss[1]);
        start = start.add(amt).mod(size);
        continue;
      }
    }
    start = start.multiply(skip.modPow(shuffles, size).subtract(BigInteger.ONE).multiply(skip.subtract(BigInteger.ONE).modInverse(size)));
    skip = skip.modPow(shuffles, size);
    return BigInteger.valueOf(2020).multiply(skip).add(start).mod(size);
  }

  void printArr(BigInteger skip, BigInteger start) {
    var res = IntStream.range(0, 10)
      .mapToObj(BigInteger::valueOf)
      .map(skip::multiply)
      .map(start::add)
      .map(n -> n.mod(BigInteger.valueOf(DECK_SIZE)))
      .map(BigInteger::intValue)
      .collect(Collectors.toList());
    for (int i = 0; i < res.size(); i++) {
      cards.set(res.get(i), i);
    }
    sopl(cards);
  }
}

File: 2019/python/advent_22.py
from common.advent_lib import *
from collections import Counter, defaultdict, deque
import collections as coll
import functools as func
import itertools as it
import operator as oper
import re
import more_itertools as it2
import numpy as np
import sympy as sym

lines = read_lines(22)
deck=deque(range(10007))
for line in lines:
    if line=="deal into new stack":
        deck.reverse()
    elif line.startswith("cut"):
        deck.rotate(-int(line.split()[-1]))
    elif line.startswith("deal with increment"):
        L=len(deck)
        skip=int(line.split()[-1])
        new_deck=deck.copy()
        i=0
        while deck:
            new_deck[i]=deck.popleft()
            i+=skip
            i%=L
        deck=new_deck


part1 = [i for i,c in enumerate(deck) if c==2019][0]
print("Part 1:", part1)

CARDS=119315717514047
SHUFFLES=101741582076661

skip=1
start=0
for line in lines[::-1]:
    if line.startswith("deal into"):
        skip = (-skip)%CARDS
        start=(CARDS-start-1)%CARDS
    elif line.startswith("deal with"):
        increment=int(line.split()[-1])
        skip = (skip*sym.mod_inverse(increment,CARDS))%CARDS
        start = (start*sym.mod_inverse(increment,CARDS))%CARDS
    elif line.startswith("cut"):
        amount=int(line.split()[-1])
        start=(start+amount)%CARDS

start *= (pow(skip, SHUFFLES, CARDS) - 1) * sym.mod_inverse(skip - 1, CARDS)
skip = pow(skip,SHUFFLES,CARDS)
part2 = (2020*skip+start)%CARDS
print("Part 2:", part2)
